### DISCUSSION TITLE
SML: SML decode 1 message offset -- parsing of first meter message is not happening

### DISCUSSION THREAD

**Discussion by HenningAust** [arendst/Tasmota#20029] [Q&A]
Hi, 

I have problem with a rather special setup. I have two Allmess Integral-V UlrtaLite Pro battery powered heat meters in my heating system with infrared serial interface. In order not to drain battery ver quick I only read data every 4h. I have to send a wake up sequence and request data and get then 1 answer telegram.
I have compiled Tasmota with serial buffer size 200, which is bigger then the 166 bytes ma meter answers / I see on the console dump:

```
21:43:00.121 wakeup start
21:43:03.093 wakeup end
21:43:03.095 wait for the meter
21:43:03.449 request data
21:43:03.528 timer
21:43:03.700 : 68 4d 4d 68 08 00 72 07 18 65 17 92 26 17 04 19 04 00 00 0c 78 07 18 65 17 04 06 0f 02 00 00 0c 
21:43:03.828 : 14 37 17 01 00 0b 2d 12 00 00 0b 3b 65 03 00 0a 5a 87 03 0a 5e 57 03 0b 61 01 03 00 04 6d 24 0e 
21:43:03.955 : f8 28 02 27 39 08 09 fd 0e 07 09 fd 0f 11 0f 00 04 47 16 
21:43:04.553 timer
21:43:05.556 timer
```

In general this works, but after the first read request all meter values stay at 0, 4 hours later I get then the values from 4 hours ago (I see this by comparing to the values shown on the display). Sometime it also looks like not the complete telegram is parsed / I get data from two different telegrams shown on my Tasmota.

I tried to figure out how sml driver works by checking the source, but it beyond my rusty C skills.
This is the script I am running, the wakeup sequence is in the S section:

```
`>D
;start, define variables
cnt=1
timer=1
hours=1
w_new=0
w_delta=0
p:w_last=0

>B
;setup sensor
->sensor53 r

>T
w_new=WAERME#w_total

>S
timer=int(time)
hours=int(timer/240) ;alle 4h
print timer %timer%
print hours %hours%
if chg[hours]>0 
then
timer=0
switch timer
case 0
print It is midnight
print wakeup start
sml(-1 1 "2400:8N1")
for cnt 1 72 1
sml(1 1 "55555555555555555555")
next
print wakeup end
print wait for the meter
delay(350)
sml(-1 1 "2400:8E1")
print request data
;sml(1 1 "6804046853FE5000A116")
;delay(100)
sml(1 1 "105BFE5916")
;case 1
;print It is a minute after midnight
;print calculating daily value
;print w_last %0w_last%
;w_delta=w_new-w_last
;w_last=w_new
;svars
;print w_new %0w_new%
;print w_delta %0w_delta%
ends
endif

;>J  
;,"w_delta":%w_delta% 

;>W
;===============
;Vortagsverbrauch:    {m} %3w_delta% kWh 

>M 1
+1,3,rE1,0,2400,WAERME,1
1,0406uuUUuuUUs@1,Total energy,kWh,w_total,0

1,0C14bcd8@100,Total volume,m³,v_total,2

1,0B2Dbcd6@10,Current power,KW,p_akt,2

1,0B3Bbcd6@1000,Current flow,m³/h,v_akt,2

1,0A5Abcd4@10,Flow temp,°C,t_flow,1

1,0A5Ebcd4@10,Return temp,°C,t_return,2

1,0B61bcd6@100,Temp diff,°C,t_diff,2

1,0227uuUU@1,Meter days,d,days_op,0
#`
```

**Comment by gemu2015** [comment]
that is normal behavior in mode 'r'.  incoming data is shifted and compared after each incoming byte.
this mode is not recommended for requested data with long intervals.
with the current driver you may decrease the serial buffer to about 1 line length

edit:

or wait until i provide a solution for this non trivial problem.

**Reply by gemu2015** [reply]
i provided a solution for this problem. for testing please use my fork: https://github.com/gemu2015/Sonoff-Tasmota

then in the above script insert this command

after line
sml(1 1 "105BFE5916")

insert
delay(100)
; this shifts the buffer x chars with decoding,  set the size of the message here as x
smls(1 x)

**Reply by HenningAust** [reply]
Hi, I will try this tomorrow and report. Though I am still don't understand whats happening under the hood here

**Reply by gemu2015** [reply]
ok, you probably should use this approach which will work with the current driver

the 'r' mode shifts data through a buffer and the decoder compares the beginning of the buffer with the decode lines while
pm() pattern match searches the complete buffer for that pattern (see docs)
r denotes raw pattern match with hex bytes 
```
>M 1
+1,3,rE1,0,2400,WAERME,1
1,pm(r0406)uuUUuuUUs@1,Total energy,kWh,w_total,0
1,pm(r0C14)bcd8@100,Total volume,m³,v_total,2
1,pm(r0B2D)bcd6@10,Current power,KW,p_akt,2
1,pm(r0B3B)bcd6@1000,Current flow,m³/h,v_akt,2
1,pm(r0A5A)bcd4@10,Flow temp,°C,t_flow,1
1,pm(r0A5E)bcd4@10,Return temp,°C,t_return,2
1,pm(r0B61)bcd6@100,Temp diff,°C,t_diff,2
1,pm(r0227)uuUU@1,Meter days,d,days_op,0
#
```

**Reply by gemu2015** [reply]
i found a regression inside pattern raw mode, please use my fork for a corrected version

i tested with your data set from above and it works now

**Reply by HenningAust** [reply]
Hi,

I pulled your code an uploaded it, it shows now 13.2.0.2 as version. Took some time to get platform.io running and figure out that upload will not work when having IR recevier on RX/TX...

it showed all values updated immediately after starting the script with your new pattern matches, need to wait for the next update at 12:00 to see what happens, but seems to be ok at the moment. 

Thx for helping me on this. I think when we have all figured out I can provide all information to put to the SML description on the website

**Reply by gemu2015** [reply]
ok, when your tests are done, i will do a PR

**Reply by HenningAust** [reply]
Mh, it still doesn't seem to parse each polled message:

sent values before and after polling do not change all, some seem to change though, it think at leat t_flow has a very small chance to not change between readings

`
15:56:26.648 MQT: tele/tasmota_0690D2/SENSOR = {"Time":"2023-11-17T15:56:26","WAERME":{"w_total":921,"v_total":182.53,"p_akt":0.60,"v_akt":0.06,"t_flow":48.8,"t_return":39.30,"t_diff":9.56,"days_op":2114}}
16:00:00.668 It is midnight
16:00:00.670 wakeup start
16:00:03.640 wakeup end
16:00:03.641 wait for the meter
16:00:03.994 request data
16:01:29.083 MQT: tele/tasmota_0690D2/STATE = {"Time":"2023-11-17T16:01:29","Uptime":"0T01:20:17","UptimeSec":4817,"Heap":10,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"H&M","BSSId":"E8:DF:70:08:49:F2","Channel":11,"Mode":"11n","RSSI":46,"Signal":-77,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:01:29.092 MQT: tele/tasmota_0690D2/SENSOR = {"Time":"2023-11-17T16:01:29","WAERME":{"w_total":921,"v_total":182.53,"p_akt":0.70,"v_akt":0.05,"t_flow":50.2,"t_return":39.30,"t_diff":9.56,"days_op":2114}}

16:21:29.108 MQT: tele/tasmota_0690D2/STATE = {"Time":"2023-11-17T16:21:29","Uptime":"0T01:40:17","UptimeSec":6017,"Heap":10,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"H&M","BSSId":"E8:DF:70:08:49:F2","Channel":11,"Mode":"11n","RSSI":46,"Signal":-77,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:21:29.117 MQT: tele/tasmota_0690D2/SENSOR = {"Time":"2023-11-17T16:21:29","WAERME":{"w_total":921,"v_total":182.53,"p_akt":0.70,"v_akt":0.05,"t_flow":50.2,"t_return":39.30,"t_diff":9.56,"days_op":2114}}
16:26:29.130 MQT: tele/tasmota_0690D2/STATE = {"Time":"2023-11-17T16:26:29","Uptime":"0T01:45:17","UptimeSec":6317,"Heap":10,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"H&M","BSSId":"E8:DF:70:08:49:F2","Channel":11,"Mode":"11n","RSSI":46,"Signal":-77,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:26:29.140 MQT: tele/tasmota_0690D2/SENSOR`

**Reply by gemu2015** [reply]
This method requires that your serial buffer is not much larger then the total message length. Otherwise you get older values which are still in front of the buffer. You may set the serial buffer size by special option, see docs

**Reply by HenningAust** [reply]
Ok I can recompile with different buffer length. How is the length of the message counted? The blanks in the dump are not counted most probably? 

I think Buffer length must no be bigger than the message by more then 2 chars to make sure none of the keywords are two times in the buffer?


I the beginning I was wondering if there is a command to empty the buffer before requesting a new reading from the meter.

**Reply by gemu2015** [reply]
you do not need to recompile! there is a simple code line in descriptor to set buffer size.
e.g. sets buffer size to 128
1,=so3,128
add this line to sml descriptor
spaces do not count.  your data set is 83 bytes, so probably choose 90

**Reply by HenningAust** [reply]
great, with this addition it looks like it works based on the data of the last 24h.

Thanks for helping figure this out.
